### DISCUSSION TITLE
how to use tailwinds @apply with @layer directive from a Svelte component

### DISCUSSION THREAD

**Discussion by opensas** [tailwindlabs/tailwindcss#5509] [Help]
### What version of Tailwind CSS are you using?

"tailwindcss": "^2.2.4"

### What build tool (or framework if it abstracts the build tool) are you using?

SvelteKit

### What version of Node.js are you using?

v14.17.6

### What browser are you using?

Firefox, chrome

### What operating system are you using?

Linux xps 5.4.0-84-generic #94~18.04.1-Ubuntu

### Reproduction repository

https://github.com/opensas/tailwind-apply

### Describe your issue

I want to use apply to define some css setting on a component, and I also want to be able to overwrite it, like this:

```svelte
<!-- CustomButton.svelte -->
<script>
    let className = '';
    export { className as class };
    export let label = 'Click me!';
</script>

<button class="custom-button {className}">{label}</button>

<style lang="postcss">
.custom-button {
    @apply bg-blue-400 font-bold text-white rounded-lg p-4;
}
</style>
```

And I want to use it like this:

```svelte
<script>
    import CustomButton from './CustomButton.svelte';
</script>

<div class="w-screen h-screen flex justify-center items-center">
    <CustomButton class="bg-red-800" label="This is my button" />
</div>
```

That is, I want to be able to override my @applied settings

The problem is that the settings from the `@apply` directives cannot be overriden by this line

```svelte
<button class="custom-button {className}">{label}</button>
```

I understand that in order to do that I have to tell tailwind to generate the corresponding css in the components layer, that is, before the utilities.

If I enter the same css directive in my app.post.css file, before the @tailwind utilities line, or using the @layer components directives it works ok:

```css
/* app.post.css */

@tailwind base;
@tailwind components;
@tailwind utilities;

/* This works ok! */
@layer components {
  .custom-button {
    @apply bg-blue-400 font-bold text-white rounded-lg p-4;
  }
}
```

So, what I want is some way to tell tailwind to add the `.custom-button` setting I defined in my `CustomButton.svelte` component to the components layer, so that it could be overriden with inline clases.

If I try to do this in my CustomButton.svelte file

```svelte
<style lang="postcss">
    @layer components {
        .custom-button {
            @apply bg-blue-400 font-bold text-white rounded-lg p-4;
        }
    }
</style>
```

I get the following error:

```
9:13:34 PM [vite] Internal server error: /home/sas/devel/apps/glas-it/apps/wingback/end-user-portal/src/routes/CustomButton.svelte?svelte&type=style&lang.css:1:1: `@layer components` is used but no matching `@tailwind components` directive is present.
  Plugin: vite:css
```

This issue is preventing me from using the @apply directive from any component.

Is there some way to achieve this?

Steps to reproduce

open: https://gitpod.io#github.com/opensas/tailwind-apply

```
npm install
npm run dev
```

click on `http://localhost:3000`

Allow popups from...

This is the result:

![image](https://user-images.githubusercontent.com/481687/133160683-2331404d-4305-4b96-a014-753bb80e14a8.png)

Expected behaviour:

I'd like to be able to define @apply commands that should go to the @component layer from a svelte component, so that I can override it with inline styles.

---

see also: this question at [SO](https://stackoverflow.com/questions/69059757/how-to-use-tailwinds-apply-with-layer-directive-from-a-svelte-component)

**Comment by nbgoodall** [comment]
Correct me if I'm wrong but this doesn't seem like an issue with Tailwind, perhaps something for the [discussions board](https://github.com/tailwindlabs/tailwindcss/discussions), or SO as you already mentioned?

**Comment by adamwathan** [comment]
Hey @opensas! The issue here is just in how tools like SvelteKit (and basically all others) process multiple CSS sources like this.

Basically what happens is they process every style block from every Svelte component independently, so if you have a global CSS file and 12 Svelte components that all have a style block, SvelteKit is running Tailwind 13 separate times that are all isolated from each other. So when you try to use `@layer components { ... }` in a Svelte component, Tailwind has no idea that there is a `@tailwind components` directive in your global CSS file because it might as well be in someone else's project on a different computer in another country, since everything is processed independently in isolation.

Only real options here are:

1. Put the custom class in your global CSS file instead of the component so that Tailwind can relocate the styles properly.
2. Don't use a custom class here at all and just use the utilities right in the markup.
3. Use the important modifier for your overrides like `!bg-red-800`

Generally I would recommend against these sort of class-based styling APIs for components because of these specificity issues, and instead do something like:

```
<CustomButton color="red" label="This is my button" />
```

...where the logic to add the right utility classes based on the selected color happens inside the component. Hope that helps!

**Reply by opensas** [reply]
Hi Adam, thanks a lot for taking the time to read and answer this.

From what I understand, the `@layer` directive is completely unusable from svelte components. I understand that Svelte processing css files independently is a serius shortcoming, as it prevents me from being able to use `@apply` to style a component and being able to override it.

Shouldn't this be considered a bug? not tailwind's nor sveltkit's bug, but a problem with the integration of both tools.

The real options you provide are workarounds, I still can't use the @layer directive from a svelte component.

I would add another workaround, don't use the `@apply` directive from components (if you want to override them) and instead use a js variable (or hardcode it), like this:

```svelte
<!-- Button.svelte -->
<script>
  let className = '';
  export { className as class };
  export let success = false;
  const style = 'bg-blue-400 font-bold text-white rounded-lg p-4'
</script>

<button class="{style} m-4 {success ? 'bg-green-800' : ''} {className}"
  ><slot>Click me!</slot></button
>
```

**Comment by PsyGik** [comment]
> Don't use a custom class here at all and just use the utilities right in the markup.

Isn't that counter productive?

Say I have an app with structure like so:

```css
/* index.css */

@tailwind base;
@tailwind components;
@tailwind utilities;

.code {
  @apply rounded overflow-hidden;
  box-shadow: 0 3px 3px rgba(0, 0, 0, 0.16), 0 3px 3px rgba(0, 0, 0, 0.23);
  border: 1px solid hsl(225, 11%, 24%);
}
.divider{
  @apply divide-y divide-gray-200 divide-dashed dark:divide-gray-700
}
```

```js
// index.js
import './index.css'
.
.

```

With my button component,

```css
/* button.css */
@layer components {
  .btn-base{
  @apply p-1 inline-flex flex-row font-medium text-gray-900 sm:p-4 dark:text-gray-100;
  /* some more custom style overrides, not from tailwind */
  }
}
```

```jsx
// button.js
import './button.css';

const Button = () =>  <button className='btn-base'/>
```

Now if I were to use this `Button` anywhere else in the app, it'll error as expected `@layer components is used but no matching @tailwind components directive is present.`  Without `@layer`, the specificity of the component style is lower than the base styles and it seems it never gets applied. 😢 

The only way I can think of fixing this is to move all component styles to `index.css`, which defeats the purpose of component specific styles. Or maybe I'm dumb and don't really understand how Tailwind works. 🤷🏽 

---

Also, probably a personal preference, but 

```jsx
<button className={styles.btnBase}/> OR <button className="btn-base"/>
``` 
looks cleaner than 

```jsx
<button className="p-1 inline-flex flex-row font-medium text-gray-900 sm:p-4 dark:text-gray-100"/>
```
